File: 3.4_heritage/heritage.cpp
/*
ID: dermotl1
LANG: C++
PROG: heritage
*/

#include <algorithm>
#include <fstream>
#include <iostream>
#include <limits>
#include <queue>
#include <utility>
#include <vector>

using namespace std;

char in[27];
int in_length = 0;
char pre[27];
int pre_length = 0;
char post[27];
int post_tracker = 0;

void solve(int in_first, int in_last, int pre_first,
           int pre_last) {  // declare ranges

  if (in_first > in_last) return;

  char root = pre[pre_first];
  int root_tracker;

  for (int i = in_first; i <= in_last; ++i) {
    if (in[i] == root) {
      root_tracker = i;
      break;
    }
  }
  int left_nodes = root_tracker - in_first;

  solve(in_first, root_tracker - 1, pre_first + 1,
        pre_first + left_nodes);  // left branch
  solve(root_tracker + 1, in_last, pre_first + left_nodes + 1,
        pre_last);  // right branch

  post[post_tracker] = root;
  post_tracker++;
}
int main() {
  ifstream fin("heritage.in");
  ofstream fout("heritage.out");

  char temp;
  while (fin.get(temp) && temp != '\n') {
    in[in_length] = temp;
    in_length++;
  }
  while (fin.get(temp) && temp != '\n') {
    pre[pre_length] = temp;
    pre_length++;
  }

  solve(0, in_length - 1, 0, pre_length - 1);

  for (int i = 0; i < post_tracker; i++) {
    fout << post[i];
  }
  fout << "\n";

  return 0;
}

File: 3.4_heritage/heritage.java
/*
ID: dermotl1
LANG: JAVA
PROG: heritage
*/

import java.io.*;
import java.lang.*;
import java.util.*;

public class heritage {

  static char in_array[] = new char[27];
  static int in_length = 0;

  static char pre_array[] = new char[27];
  static int pre_length = 0;

  static char post_array[] = new char[27];
  static int post_tracker = 0;

  public static void solve(int in_first, int in_last, int pre_first,
      int pre_last) {
    if (in_first > in_last)
      return;

    char root = pre_array[pre_first];
    int root_tracker = 0;

    for (int i = in_first; i <= in_last; ++i) {
      if (in_array[i] == root) {
        root_tracker = i;
        break;
      }
    }

    int left_nodes = root_tracker - in_first;

    solve(in_first, root_tracker - 1, pre_first + 1,
        pre_first + left_nodes);
    solve(root_tracker + 1, in_last, pre_first + left_nodes + 1,
        pre_last);

    post_array[post_tracker] = root;
    post_tracker++;
  }

  public static void main(String args[]) throws Exception {
    Scanner in = new Scanner(new File("heritage.in"));
    PrintWriter out = new PrintWriter(new File("heritage.out"));

    char temp;
    String in_string = in.nextLine();
    String pre_string = in.nextLine();

    if (in_string != null) {
      in_array = in_string.toCharArray();
      in_length = in_string.length();
    }

    if (pre_string != null) {
      pre_array = pre_string.toCharArray();
      pre_length = pre_string.length();
    }

    solve(0, in_length - 1, 0, pre_length - 1);

    for (int i = 0; i < post_tracker; i++) {
      out.print(post_array[i]);
    }
    out.println();

    in.close();
    out.close();
  }
}